modules/appeals_api/lib/appeals_api/stats_report.rb

```diff
@@ -0,0 +1,147 @@
+# frozen_string_literal: true
+
+module AppealsApi
+  class StatsReport
+    DATE_FORMAT = '%b%e, %Y'
+
+    STATUS_TRANSITION_PAIRS = [
+      %w[processing submitted], %w[submitted complete], %w[processing success], %w[error success]
+    ].freeze
+
+    STALLED_THRESHOLD = 2.months
+
+    def initialize(date_from, date_to)
+      @date_from = date_from.to_date
+      @date_to = date_to.to_date
+    end
+
+    def text
+      <<~REPORT
+        Appeals stats report
+        #{date_from.strftime(DATE_FORMAT)} to #{date_to.strftime(DATE_FORMAT)}
+        ===
+
+        Higher Level Reviews
+        ---
+        ### Transition times
+
+        #{formatted_stats(AppealsApi::HigherLevelReview)}
+        ### Stalled records
+
+        #{formatted_stalled(AppealsApi::HigherLevelReview)}
+
+
+        Notice of Disagreements
+        ---
+        ### Transition times
+
+        #{formatted_stats(AppealsApi::NoticeOfDisagreement)}
+        ### Stalled records
+
+        #{formatted_stalled(AppealsApi::NoticeOfDisagreement)}
+
+
+        Supplemental Claims
+        ---
+        ### Transition times
+
+        #{formatted_stats(AppealsApi::SupplementalClaim)}
+        ### Stalled records
+
+        #{formatted_stalled(AppealsApi::SupplementalClaim)}
+      REPORT
+    end
+
+    private
+
+    attr_accessor :date_from, :date_to
+
+    def status_update_records(statusable_type, status_from, status_to)
+      @records ||= {}
+      @records[statusable_type] ||= {}
+      @records[statusable_type][status_from] ||= {}
+      @records[statusable_type][status_from][status_to] ||=
+        begin
+          records = AppealsApi::StatusUpdate.where(
+            from: status_from,
+            to: status_to,
+            statusable_type: statusable_type,
+            status_update_time: date_from..date_to
+          ).order(:statusable_id)
+          previous_records = AppealsApi::StatusUpdate.where(
+            to: status_from,
+            statusable_id: records.pluck(:statusable_id),
+            statusable_type: statusable_type
+          ).order(:statusable_id)
+          records.zip(previous_records)
```

Comment:
Double check that the records.pluck & this records.zip doesn't send two queries to the database. If it does, you should be able to first grab all the records, then in `previous_records` map the `id` out.

Also, we do a lot more status updates than we do created appeal records (each appeal record will have multiple status updates) so let's only `select` the columns we're going to use (that way we're not pulling in the error codes & details when we don't need to).

Reply:
Remind me to double check the database for this before you work the below - if we've already removed the duplicates then we don't have to worry about filtering things out:

I think we removed them already, but one thing to watch out for is that at one point we were storing updates even when the `from` and `to` were the same. I don't think `records` will have an issue, but `previous_records` might - perhaps a good failsafe would be to update the query to make sure to only return records where the `from` and `to` aren't the same.

Reply:
Ah ok, thank you! I added the filter before I saw this comment, but I think we recently updated things so that an error state can follow another error state (as long as they have different code/status) - maybe that doesn't matter (since it's still a different kind of state), but I left in the filter here